qlib/kernel/arch/aarch64/syscall_entry.s

```diff
@@ -7,6 +7,52 @@
 // TODO rename this file
 __vsyscall_page:
 CopyPageUnsafe:
+	ldp	x2, x3, [x1]
+	ldp	x4, x5, [x1, #16]
+	ldp	x6, x7, [x1, #32]
+	ldp	x8, x9, [x1, #48]
+	ldp	x10, x11, [x1, #64]
+	ldp	x12, x13, [x1, #80]
+	ldp	x14, x15, [x1, #96]
+	ldp	x16, x17, [x1, #112]
+
+	add	x0, x0, #256
+	add	x1, x1, #128
+1:
+    // PAGE_SIZE
+	tst	x0, #(0x1000 - 1)
```

Comment:
if the supplied address is not 4k aligned, this loop will not terminate properly (perhaps it would end at all).  I would suggest either 1) explicitly check for alignment in the wrapper function or 2) use `cmp` instead of `tst` for robustness.

Also I think the loop is unrolled too much? Perhaps 64 bytes in one iteration is enough.

Otherwise LGTM.

Reply:
I agree with @shrik3; the alignment test should be done beforehand.

Reply:
Do we still have to do the alignment? It seems all the callers of the CopyPageUnsafe did made sure the source is page alignment.

Reply:
> It seems all the callers of the CopyPageUnsafe did made sure the source is page alignment.

I'm fine with this, as long as we keep it in mind. A code comment would be nice, e.g.:

```
/// UNSAFE CODE! the caller must make sure the to and from address are 4k aligned.
```

Reply:
done

Reply:
done